server/storage/backend/batch_tx.go

```diff
@@ -73,7 +73,8 @@ type batchTx struct {
 	tx      *bolt.Tx
 	backend *backend
 
-	pending int
+	pending                 int
+	pendingDeleteOperations int
```

Comment:
It's not clear for me why we track deletions on the `batchTx` instead on the `batchTxBuffered`. The issue we originally discussed is about buffer in `batchTxBuffered` not representing deletes. Is there any logic in the `batchTx` I missed that would also require handling this?

Reply:
It's natural and easy to read to track two metrics (`pending***`) in one place. It isn't a big problem.

Reply:
Don't agree, but give me some time to read the code again.

Reply:
I worry about separation of concern in the batch_tx.go package. Fact that `batchTx` and `batchTxBuffered` are used together doesn't mean we should entangle them. For example for some reason precommit hooks are executed by of `batchTxBuffered` and not `batchTx`. Why? I don't know, maybe hook call location needed to be ordered in relation to buffer, however It's hard to know now without deep dive.

I agree that this isn't a big problem now, simplifying the transaction logic would be another large effort that we can postpone for later.